File: python/sparc/streaming.py
"""
Streaming / memory-efficient processing for large FASTQ files.
"""

from pathlib import Path
from typing import Callable, Iterator, Optional, Union
from dataclasses import dataclass

try:
    from sparc._sparc_py import FastqParser, FastqRecord
    _RUST_AVAILABLE = True
except ImportError:
    _RUST_AVAILABLE = False


@dataclass
class StreamStats:
    """Statistics from streaming processing."""
    total_records: int = 0
    processed_records: int = 0
    chunks_processed: int = 0


class StreamingProcessor:
    """
    Memory-efficient streaming processor for large FASTQ files.

    Processes data in configurable chunk sizes to limit memory usage.

    Parameters
    ----------
    chunk_size : int
        Number of records per chunk (default: 100000)
    max_memory_mb : int
        Approximate memory limit in MB (default: 4096)
    """

    def __init__(self, chunk_size: int = 100_000, max_memory_mb: int = 4096):
        self.chunk_size = chunk_size
        self.max_memory_mb = max_memory_mb

    def process_fastq(
        self,
        path: Union[str, Path],
        callback: Callable[[list], None],
    ) -> StreamStats:
        """
        Process a FASTQ file in chunks.

        Parameters
        ----------
        path : str or Path
            Path to FASTQ file
        callback : callable
            Function called with each chunk of records

        Returns
        -------
        StreamStats
            Processing statistics
        """
        if not _RUST_AVAILABLE:
            raise ImportError("Rust bindings not available. Install with: pip install sparc")

        stats = StreamStats()
        parser = FastqParser(str(path))
        chunk = []

        for record in parser:
            chunk.append(record)
            stats.total_records += 1

            if len(chunk) >= self.chunk_size:
                callback(chunk)
                stats.chunks_processed += 1
                stats.processed_records += len(chunk)
                chunk = []

        if chunk:
            callback(chunk)
            stats.chunks_processed += 1
            stats.processed_records += len(chunk)

        return stats

    def iter_chunks(
        self,
        path: Union[str, Path],
    ) -> Iterator[list]:
        """
        Iterate over chunks of FASTQ records.

        Parameters
        ----------
        path : str or Path
            Path to FASTQ file

        Yields
        ------
        list
            Chunk of FastqRecord objects
        """
        if not _RUST_AVAILABLE:
            raise ImportError("Rust bindings not available. Install with: pip install sparc")

        parser = FastqParser(str(path))
        chunk = []

        for record in parser:
            chunk.append(record)
            if len(chunk) >= self.chunk_size:
                yield chunk
                chunk = []

        if chunk:
            yield chunk

File: crates/sparc-core/src/streaming.rs
//! Streaming/chunked processing for large datasets

use crate::fastq::{FastqParser, FastqRecord};
use crate::Result;
use std::path::Path;

/// Configuration for streaming processing
pub struct StreamConfig {
    /// Number of records per chunk
    pub chunk_size: usize,
    /// Maximum memory usage in bytes (advisory)
    pub max_memory: usize,
}

impl Default for StreamConfig {
    fn default() -> Self {
        Self {
            chunk_size: 1_000_000,
            max_memory: 4 * 1024 * 1024 * 1024, // 4GB
        }
    }
}

/// Statistics from streaming processing
#[derive(Debug, Default)]
pub struct StreamStats {
    pub total_records: u64,
    pub chunks_processed: u64,
}

/// Streaming FASTQ processor that processes records in fixed-size chunks
pub struct StreamingProcessor {
    config: StreamConfig,
}

impl StreamingProcessor {
    pub fn new(config: StreamConfig) -> Self {
        Self { config }
    }

    pub fn with_chunk_size(chunk_size: usize) -> Self {
        Self {
            config: StreamConfig {
                chunk_size,
                ..Default::default()
            },
        }
    }

    /// Process FASTQ file in chunks, calling the callback for each chunk
    pub fn process_fastq<P, F>(&self, path: P, mut callback: F) -> Result<StreamStats>
    where
        P: AsRef<Path>,
        F: FnMut(&[FastqRecord]) -> Result<()>,
    {
        let mut parser = FastqParser::open(path)?;
        let mut stats = StreamStats::default();
        let mut chunk = Vec::with_capacity(self.config.chunk_size);

        loop {
            chunk.clear();
            for _ in 0..self.config.chunk_size {
                match parser.next() {
                    Some(Ok(record)) => chunk.push(record),
                    Some(Err(e)) => return Err(e),
                    None => break,
                }
            }
            if chunk.is_empty() {
                break;
            }
            stats.total_records += chunk.len() as u64;
            stats.chunks_processed += 1;
            callback(&chunk)?;
        }
        Ok(stats)
    }

    /// Process FASTQ file in parallel chunks using rayon
    pub fn process_fastq_parallel<P, F, T>(&self, path: P, f: F) -> Result<Vec<T>>
    where
        P: AsRef<Path>,
        F: Fn(&FastqRecord) -> T + Send + Sync,
        T: Send,
    {
        use rayon::prelude::*;

        let mut parser = FastqParser::open(path)?;
        let records = parser.read_all()?;

        let results: Vec<T> = records
            .par_chunks(self.config.chunk_size)
            .flat_map(|chunk| chunk.iter().map(&f).collect::<Vec<_>>())
            .collect();

        Ok(results)
    }
}
